src/identifier.rs

```diff
@@ -70,18 +70,83 @@ use crate::alloc::alloc::{alloc, dealloc, Layout};
 #[cfg(no_from_ne_bytes)]
 use crate::backport::FromNeBytes;
 use core::mem;
-use core::num::{NonZeroU64, NonZeroUsize};
+use core::num::NonZeroUsize;
 use core::ptr;
+use core::ptr::NonNull;
 use core::slice;
 use core::str;
 
 #[repr(transparent)]
 pub(crate) struct Identifier {
-    repr: NonZeroU64,
+    repr: Repr,
+}
+
+use crate::identifier::repr::Repr;
+
+mod repr {
+    use core::ptr::NonNull;
+
+    #[cfg(target_pointer_width = "64")]
+    #[derive(PartialEq, Eq, Clone, Copy)]
+    pub struct Repr {
+        ptr: NonNull<u8>,
+    }
+
+    #[cfg(target_pointer_width = "32")]
+    #[derive(PartialEq, Eq, Clone, Copy)]
+    pub struct Repr {
+        ptr: NonNull<u8>,
+        remainder: usize,
+    }
+
+    impl Repr {
+        pub fn as_ptr(self) -> *const u8 {
+            self.ptr.as_ptr()
+        }
+
+        #[cfg(target_pointer_width = "32")]
+        pub const fn from_ptr(ptr: NonNull<u8>) -> Self {
+            Repr {
+                remainder: 0,
+                ptr: ptr,
+            }
+        }
+
+        #[cfg(target_pointer_width = "32")]
+        pub unsafe fn from_inline(bytes: u64) -> Repr {
+            Repr {
+                ptr: unsafe { NonNull::new_unchecked(bytes as usize as *mut u8) },
+                remainder: (bytes >> 32) as usize,
```

Comment:
Due to the lack of as_u64, I don't see anywhere remainder is ever read. Is it actually a 64 bit value or is it just a isize?

Reply:
It is read both in the `is_inline` method, and a pointer to it is produced by `inline_as_str`. As I said in my initial comment, all this juggling to support 32-bit platforms makes me uncomfortable and I think the `MaybeUninit` implementation would be simpler, but whether it is viable at all is up to the maintainer.

Reply:
https://doi.org/10.1287/orsc.2018.1277

Reply:
Oh, that was exceedingly unclear. Maybe that cast chain could be moved to an impl on Repr?